diff --git a/src/Plugin/EntityBrowser/Widget/View.php b/src/Plugin/EntityBrowser/Widget/View.php
index 605f782..76d55a0 100644
--- a/src/Plugin/EntityBrowser/Widget/View.php
+++ b/src/Plugin/EntityBrowser/Widget/View.php
@@ -143,9 +143,15 @@ class View extends WidgetBase implements ContainerFactoryPluginInterface {
       }
     }

-    $form['view']['view'] = [
-      '#markup' => \Drupal::service('renderer')->render($form['view']['view']),
-    ];
+    try {
+      $form['view']['view'] = [
+        '#markup' => \Drupal::service('renderer')->render($form['view']['view']),
+      ];
+    } catch (\Exception $exception) { // Yep, I know it's bad, don't care
+      $form['view']['view'] = [
+        '#markup' => \Drupal::service('renderer')->renderRoot($form['view']['view']),
+      ];
+    }

     return $form;
   }

